server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadata.java

```diff
@@ -101,11 +159,76 @@ public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         DesiredNodesMetadata that = (DesiredNodesMetadata) o;
-        return Objects.equals(latestDesiredNodes, that.latestDesiredNodes);
+        return Objects.equals(latestDesiredNodes, that.latestDesiredNodes)
+            && Objects.equals(clusterMembers, that.clusterMembers)
+            && Objects.equals(notClusterMembers, that.notClusterMembers);
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(latestDesiredNodes);
+        return Objects.hash(latestDesiredNodes, clusterMembers, notClusterMembers);
+    }
+
+    @Override
+    public String toString() {
+        return "DesiredNodesMetadata{"
+            + "latestDesiredNodes="
+            + latestDesiredNodes
+            + ", clusterMembers="
+            + clusterMembers
+            + ", notClusterMembers="
+            + notClusterMembers
+            + '}';
+    }
+
+    public static class Builder {
+        private final Logger logger = LogManager.getLogger(Builder.class);
+
+        private DesiredNodes desiredNodes;
+        private final Set<DesiredNode> clusterMembers;
+
+        public Builder(DesiredNodesMetadata desiredNodesMetadata) {
```

Comment:
nit: We often add a `builder()` method to the outer class. I like that, since it make the client code slightly shorter and reads slightly better I think.